### ISSUE TITLE
consider updating libutp dependency to 2b364cb

### ISSUE THREAD

**Issue by GaryElshaw** [transmission/transmission#2996]
Libutp is 12 years old. May 2018 is the most recent, and is surely better than using a 12 year old version. https://github.com/bittorrent/libutp.

@ckerr @mikedld @livings124 I downloaded the latest version and tried to build it for Transmission by altering the current cmakefile. A number of files have changed name, and i've included the alterations to the current makefile, and had my fingers crossed that i was doing the right thing.

These file names no longer exist:

1. utp.cpp
2. utypes.h

cmakefile changes i made are utp_api.cpp and utp_types.h:

"include_directories(.)

add_library(${PROJECT_NAME} STATIC
    utp_api.cpp
    utp_utils.cpp
    ${${PROJECT_NAME}_ADD_SOURCES}
)

install(TARGETS ${PROJECT_NAME} DESTINATION lib)
install(FILES utp.h utp_types.h DESTINATION include/libutp)"

Here's the output during the build process:

```
1 warning generated.
[ 23%] Building CXX object libtransmission/CMakeFiles/transmission.dir/natpmp.cc.o
[ 23%] Building CXX object libtransmission/CMakeFiles/transmission.dir/net.cc.o
/Users/gary/Documents/GitHub/transmission/libtransmission/net.cc:437:42: error: use of undeclared identifier 'UTP_Create'
        struct UTPSocket* const socket = UTP_Create(tr_utpSendTo, session, (struct sockaddr*)&ss, sslen);
                                         ^
/Users/gary/Documents/GitHub/transmission/libtransmission/net.cc:461:9: error: use of undeclared identifier 'UTP_Close'
        UTP_Close(socket.handle.utp);
        ^
2 errors generated.
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/net.cc.o] Error 1
make[1]: *** [libtransmission/CMakeFiles/transmission.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by mikedld** [comment]
An attempt that I made at one point could be found in the [trac5888-utorrent-3.4-utp](https://github.com/transmission/transmission/compare/trac5888-utorrent-3.4-utp) branch. As seen from its name, it was first reported as [TRAC-5888](https://trac.transmissionbt.com/ticket/5888) 7 years ago.

**Comment by GScottElshaw** [comment]
I'll keep bashing away at it, but i'm curious where and what Transmission _needs_ from the src.

**Comment by ckerr** [comment]
> I'll keep bashing away at it, but i'm curious where and what Transmission needs from the src.

I don't understand the question; could you rephrase it?

**Comment by ckerr** [comment]
@mikedld any lessons learned from trac5888-utorrent-3.4-utp? -- Why was it abandoned, were there difficulties in the upgrade?

**Comment by mikedld** [comment]
Given that the API has changed, I wasn't exactly sure that what I did there was good enough, and there was nobody around to review (sigh), that's why it wasn't merged, despite working for me locally.

**Comment by GScottElshaw** [comment]
I tried a simple 'make' and got a couple of warnings.

`g++ -Wall -DPOSIX -g -fno-exceptions -O3 -fPIC -fno-rtti -Wno-sign-compare -fpermissive   -c -o utp_internal.o utp_internal.cpp
utp_internal.cpp:157:19: warning: unused variable 'flagnames' [-Wunused-const-variable]
static const cstr flagnames[] = {
                  ^
utp_internal.cpp:172:19: warning: unused variable 'statenames' [-Wunused-const-variable]
static const cstr statenames[] = {
                  ^
2 warnings generated.`

The resulting files:

![SCR-20220429-jag](https://user-images.githubusercontent.com/69029666/165873994-7e62f7d6-502c-4a5c-9b93-29e99ff6417b.png)

![SCR-20220429-jar](https://user-images.githubusercontent.com/69029666/165874042-57a350c7-87ff-4adc-a41e-a2acdcffb44d.png)

Any guidance on how this can be integrated into Transmission and tested locally?

**Comment by ckerr** [comment]
> Given that the API has changed, I wasn't exactly sure that what I did there was good enough, and there was nobody around to review (sigh), that's why it wasn't merged, despite working for me locally.

Not a problem anymore. I'll trade you a review for Sanity Windows CI asset uploads :pray:

**Comment by mikedld** [comment]
I have rebased the aforementioned [trac5888-utorrent-3.4-utp](https://github.com/transmission/transmission/compare/trac5888-utorrent-3.4-utp) branch in case anyone wants to try it out / review / improve on.

**Comment by GaryElshaw** [comment]
Thanks for doing this Mike. I've built and run it locally for a few days and it seems to be running well. I'm still seeing errors in the debug message log like i was in #3110, but also some hints about where some of the issues might be? Note the references to peer-io and peer-mgr.

![Untitled](https://user-images.githubusercontent.com/69029666/175331796-a1aed282-49a8-4bab-82e0-68f14561bf1a.png)


![SCR-20220624-48f](https://user-images.githubusercontent.com/69029666/175331841-d72298a6-7058-4b00-9308-fb771f76cf2c.png)

Of course i tried compiling with the current libutp repo using yours as a basis. It felt kinda close.

[UTP-build-errors.txt](https://github.com/transmission/transmission/files/8968546/UTP-build-errors.txt)

**Comment by mikedld** [comment]
According to https://github.com/bittorrent/libutp/blob/2b364cbb0650bdab64a5de2abb4518f9f228ec44/utp.h#L59-L63, 2 means `UTP_ETIMEDOUT`. This, as well as the other two error codes, don't seem that interesting, so maybe we shouldn't log them at all, or at least lower the level to trace? @ckerr

**Comment by GaryElshaw** [comment]
> According to https://github.com/bittorrent/libutp/blob/2b364cbb0650bdab64a5de2abb4518f9f228ec44/utp.h#L59-L63, 2 means `UTP_ETIMEDOUT`. This, as well as the other two error codes, don't seem that interesting, so maybe we shouldn't log them at all, or at least lower the level to trace?

Those references to peer-io and peer-mgr are of no relevance or importance?

**Comment by mikedld** [comment]
Those are the places where Transmission logs the messages, not where libutp issues them, so not really helpful. But in libutp there're only 4 places where "on error" callback is called, so...

**Comment by ckerr** [comment]
@mikedld this seems to be working alright but don't see a PR for it. Is this still in draft form, or do you want this to go into main?

**Comment by mikedld** [comment]
@ckerr, was kinda waiting for some preliminary feedback but if you think the approach is sound I'll open one.

**Comment by ckerr** [comment]
> was kinda waiting for some preliminary feedback

I've been testing with it for about four days and it seems to WfM. I should've said so earlier; sorry that I didn't.
